Check email test errors for javax.mail.AuthenticationFailedException

dpsutton opened the pull request:
fixes #23918 (and with #23925 fixes #19662)

Authing with Office365 led to a less-than-great user experience. If you
type in the wrong username or password, we would display an unhelpful
message:

> Sorry, something went wrong. Please try again. Error::Exception
reading response.Read timed out.

This is an error from an inner nested exception which looks like:
```clojure
(javax.mail.AuthenticationFailedException.
 "" ;; Office365 returns auth exception with no message so we only saw "Read timed out" prior
 (javax.mail.MessagingException.
  "Exception reading response"
  (java.net.SocketTimeoutException. "Read timed out")))
```

We didn't get any message from the AuthenticationFailedException so our
humanization strategy only went on the underlying "Read timed out". Now
we can check against the error class and use that information.


#### Before

<img width="1038" alt="image" src="https://user-images.githubusercontent.com/6377293/178827855-fc6f533d-a6f0-41a2-8dd5-14d342dad88c.png">

#### After
<img width="1207" alt="image" src="https://user-images.githubusercontent.com/6377293/178827886-6d36a95f-bfe0-4073-a776-f5a9d44c036a.png">


#### Recreate:

give some invalid Office365 credentials:
- smtp host: smtp.office365.com
- smtp port: 587
- username: make up anything, I used `some@person.com`
- password: the world is your oyster here

Attempt to save those credentials. Before this change it seems like perhaps we don't work well with office365 (particularly if you don't know you mistyped your password or if you are hitting #23919 (which sends the obfuscated password on edits rather than what you think is the saved password)). Before this change you get a cryptic timeout error. After this change you get highlighted username and password fields.

dpsutton committed:
Check email test errors for javax.mail.AuthenticationFailedException

Authing with Office365 led to a less-than-great user experience. If you
type in the wrong username or password, we would display an unhelpful
message:

> Sorry, something went wrong. Please try again. Error::Exception
reading response.Read timed out.

This is an error from an inner nested exception which looks like:
```clojure
(javax.mail.AuthenticationFailedException.
 "" ;; Office365 returns auth exception with no message so we only saw "Read timed out" prior
 (javax.mail.MessagingException.
  "Exception reading response"
  (java.net.SocketTimeoutException. "Read timed out")))
```

We didn't get any message from the AuthenticationFailedException so our
humanization strategy only went on the underlying "Read timed out". Now
we can check against the error class and use that information.

camsaul commented on src/metabase/api/email.clj:
```diff
@@ -31,9 +31,16 @@
-      (condp re-find message
+          exceptions  (u/full-exception-chain error)
+          message     (str/join ": " (map ex-message exceptions))
+          check       (fn check [regex|exception-class [m es]]
```
I have nothing against using pipes in symbol names but I have no idea what this is supposed to mean. Is `regex|exception-class` a regex pattern that you pipe the name of `exception-class` into to see if it matches something? That's my best guess. Maybe this is worth a comment

camsaul replied:
Oh I get it now. It means `regex-or-exception-class`. I only figured it out by reading the code which I think is a sign that it's not the greatest variable name. I'm generally a fan of fancy characters in symbol names because we can -- I've put `₁` and `°` in names in the codebase in the past but eventually removed them after other people whined about them -- but I think we should make sure their meaning is obvious.

I think `x|y` is a neat convention but if this is the only place we're using it it's not much of a convention. So maybe we should hate fun and just call this `regex-or-exception-class`. Just so no one else gets super confused.

camsaul replied:
Other observation: `check` is a little bit terse. Check _what_? `m` and `es` are cryptic. Is `m` supposed to be a message? 

I probably sound like a broken record but code is write-once, read-many so making these variable names a little longer makes the code a lot easier to understand

dpsutton replied:
ha sorry. I've seen that as a standard convention in the past. I'll rename it so its not confusing.

dpsutton replied:
updated. thanks for the pointers.

camsaul reviewed (approved):
Overall this LGTM but I spent 90% of my time reviewing this PR trying to decipher the variable names to figure out what they meant. Do with that information what you will.

dpsutton committed:
Little bit of clarity

better function name (`check` -> `match-error`), better arg name (
`regex|exception-class` -> `regex-or-exception-class`), full argument
names `m` -> `message` and `es` -> `exceptions`
